### PR TITLE
feat(core) deno_core::extension! macro to simplify extension registration

### PR DESCRIPTION
**Pull request by mmastrac**
This implements two new macros for extension registration to reduce a lot of the boilerplate and open the door for some future improvements:

 * `deno_core::ops!` registers a block of `#[op]`s, optionally with type parameters.
 * `deno_core::extension!` is used to register an extension, and creates two methods that can be used at runtime/snapshot generation time: `init_ops` and `init_ops_and_esm`.

An extension may look something like this:

```
deno_core::extension!(deno_webstorage,
  deps = [ deno_webidl ],
  ops = deno_ops,
  esm = [ "01_webstorage.js" ],
  config = {
    origin_storage_dir: Option<PathBuf>
  },
  state = |state, origin_storage_dir| {
    // ...
  },
);
```

### CODE REVIEW

File: cli/lsp/tsc.rs
Comment (bartlomieju) [inline cli/lsp/tsc.rs:2851]: Sweet! This is exactly what I had in mind. Not sure if `config` is the name I would go with, but this is the exact idea

File: cli/tsc/mod.rs
Comment (bartlomieju) [inline cli/tsc/mod.rs:832]: Okay I don't understand this bit... Does `deno_core::ops!()` macro expand to `deno_ops` function?

File: bench_util/benches/op_baseline.rs
Comment (bartlomieju) [inline bench_util/benches/op_baseline.rs:12]: Question: is trailing comma required in the op list?

File: core/modules.rs
Comment (bartlomieju) [inline core/modules.rs:1991]: Could `deno_core::ops!` macro be called directly inside `.ops()`?

File: ext/node/lib.rs
Comment (bartlomieju) [inline ext/node/lib.rs:361]: I believe it's a TODO at this point, but we should be able to specify both `esm()` and `esm_entry_point()` in the `extension!` macro right?

File: ext/web/benches/timers_ops.rs
Comment (bartlomieju) [inline ext/web/benches/timers_ops.rs:25]: I'm thinking we should leave function names as `init_ops_and_esm` and `init_ops`. Thoughts on that?

File: core/extensions.rs
Comment (bartlomieju) [inline core/extensions.rs:109]: Would it be possible to use a static slice here?

File: core/extensions.rs
Comment (bartlomieju) [inline core/extensions.rs:131]: I'm inclined to say that it's a cruft that hasn't been cleaned up that we need to join two vectors of ops before defining them. I think it's fine we add it for now, but we should strive to remove it by reorganization of crates that join op vectors.

File: cli/ops/testing.rs
Comment (bartlomieju) [inline cli/ops/testing.rs:47]: @littledivy suggested to me that maybe we should have a single struct instead of `config` and then `state` fn would always receive either 1 (`state`) or 2 (`state` and config struct) arguments. Any thought @mmastrac?

File: core/extensions.rs
Comment (bartlomieju) [reply core/extensions.rs:109]: It's fine to do a vector in the first pass, just asking for a potential followup question

File: cli/tsc/mod.rs
Comment (mmastrac) [reply cli/tsc/mod.rs:832]: Yep. In this case when I fold the ops into the extension, that will disappear.

File: bench_util/benches/op_baseline.rs
Comment (mmastrac) [reply bench_util/benches/op_baseline.rs:12]: It doesn't need to be, but before I finish I'll make sure that all the macro rules support optional trailing commas (and clean up any trailing commas on single lines that look weird like this).

File: core/modules.rs
Comment (mmastrac) [reply core/modules.rs:1991]: I believe so. I was holding off on this for a bit just to get everything into the `extension!` macro first, but there's no reason why it can't all just be in a single macro.

File: ext/web/benches/timers_ops.rs
Comment (mmastrac) [reply ext/web/benches/timers_ops.rs:25]: That's totally fine as well. I can circle back and update all those names.

File: core/extensions.rs
Comment (mmastrac) [reply core/extensions.rs:131]: Great, that makes things a lot easier. I was trying to maintain the rough structure of the ops from before but I can clean this up.

File: cli/ops/testing.rs
Comment (mmastrac) [reply cli/ops/testing.rs:47]: That would work as well. I actually had that in an earlier pass, but I was worried that having to prefix everything w/`cfg.` added a lot of noise. I'll circle back at the end on this one as well.

File: cli/build.rs
Comment (aapoalas) [inline cli/build.rs:95]: "we are will"

File: core/extensions.rs
Comment (bartlomieju) [reply core/extensions.rs:109]: That would be great to see - even with all the improvements from #18080 there are still visible stack in the flamegraph related to initializing ops. Reducing clones and using static slices should help with that 👍

File: core/extensions.rs
Comment (littledivy) [inline core/extensions.rs:118]: maybe have something like this to avoid .push() calls...and so that its possible to convert this to a static array later.

```suggestion
      vec![
        $(
          $( #[ $m ] )*
          $( $op )::+ :: decl $( :: <$op_param> )? (),
        )+
      ]
    }
```

File: core/extensions.rs
Comment (littledivy) [inline core/extensions.rs:159]: this is sophisticated. we need to have expansion tests like in `deno_ops`, see https://github.com/eupn/macrotest

File: core/extensions.rs
Comment (littledivy) [inline core/extensions.rs:281]: nit: s/init_ops/build_ops

Comment (bartlomieju) [reply core/extensions.rs:281]: I already did this rename in a previous PR, so I would suggest to keep it as is to avoid more changes.

**Comment by bartlomieju** [comment]
Looks great @mmastrac. I think Divy's suggestions are valid - once we address them this should be good to merge.

**Review by ry — COMMENTED** [review]
I think we should be wary of introducing too much meta-programming. This patch is a net reduction of ~150 lines. But it introduces a lot of complex in macro programming. Because there's no runtime benefit, it's not obvious to me this is a clear win.

File: core/extensions.rs
Comment (mmastrac) [reply core/extensions.rs:118]: Fixed, thanks!

**Comment by mmastrac** [comment]
> @bartlomieju The static ops list can done without this macro too.
> 
> Moreover, there are places where static array won't work.
> 
> https://github.com/denoland/deno/blob/e6fe1630f0e6cbc16ccfda5a58cbe8ecdfea2694/core/runtime.rs#L765
> 
> Places where the list is mutated or collected into a new Vec...so I think we should first figure out `&'static [OpDecl]`.

The `&'static [OpDecl]` problem is pretty hairy to tackle without doing this macro and cleanup first, I think.

Barring `const` OpDecls being available, we have three options:

 * Create a large allocation at boot time to hold all the OpDecls, and leak that allocation to create a static lifetime reference
 * Use #[ctor] to initialize the ops before rust's main starts0
 * Use OnceCell to initialize a static mut array on first access

I'd lean towards the first one, but regardless of which option we choose, it'll require a massive refactoring across the codebase unless we have all the `OpDecl`s handled in one uniform way as we do with this PR. As all the "fast" extensions in the snapshot now have a common implementation, it would be trivial for us to interrogate the size of the `OpDecl` list in a `const` manner, allocate a single chunk of memory for all of them (either at runtime or in the compiler) and fill that chunk of memory in the most efficient way possible.

Without this PR landing in every case, we'd basically be making changes to ~20+ op lists across the whole codebase.

Either way, having a macro that's not perfect and exactly what we'll want six months from now is better than nothing, as iterating and experimenting becomes 20x cheaper.

**Review by littledivy — COMMENTED** [review]
> Barring const OpDecls being available, we have three options:

I'm not sure I understand. Extensions holding their individual `&'static mut [OpDecl]` could just work too. 

> Without this PR landing in every case, we'd basically be making changes to ~20+ op lists across the whole codebase.

Good point.

File: core/extensions.rs
Comment (littledivy) [reply core/extensions.rs:159]: Oh umm yeah we don't use nightly. Let's just have https://github.com/dtolnay/trybuild tests, wdyt?

**Comment by mmastrac** [comment]
> I'm not sure I understand. Extensions holding their individual &'static mut [OpDecl] could just work too.

`op_foo::decl()` isn't `const`, so we can't easily create static slices. For X extensions we'd have to leak X `Box`es, run X `#[ctor]` macros, or have X `OnceCell`-protected arrays.

File: core/extensions.rs
Comment (bartlomieju) [inline core/extensions.rs:109]: Could you change it to use `vec![]` similarly to how it's done in `with_ops` (ie. avoiding calls to `push`) that should help with superfluous allocations

Comment (mmastrac) [reply core/extensions.rs:109]: Good call, will update this one as well.

**Review by bartlomieju — APPROVED** [review]
LGTM, fantastic first PR. 

Please update the commit message before landing (ie. remove mention of `ops_bundle!`)

**Comment by littledivy** [comment]
> op_foo::decl() isn't const, so we can't easily create static slices. For X extensions we'd have to leak X Boxes, run X #[ctor] macros, or have X OnceCell-protected arrays.

Right. decl() shouldn't be hard to make `const` though. https://github.com/denoland/deno/blob/fe07f6c3f890aacc4161fb9df771d1f221b6bd9a/core/extensions.rs#L48-L56

It's just the fast_fn that can be converted, it can be made a fixed sized trait.

**Review by littledivy — APPROVED** [review]
LGTM 2 - Great PR

File: core/extensions.rs
Comment (mmastrac) [reply core/extensions.rs:156]: I like the idea of using `macrotest` more than `trybuild` TBH but I think I'd like to split this off into a new PR and experiment a bit. The `trybuild` tests don't give much signal beyond the codebase compiling and I think there's some value in both seeing the macro output and failing when things change like your nice `ops` tests.

**Comment by mmastrac** [comment]
> It's just the fast_fn that can be converted, it can be made a fixed sized trait.

This would be great to have. I think with that we could probably move most of the extension setup to compile-time which would be pretty sweet. I wasn't able to get a good handle on what that particular change would require as it went into some of the magic in `rusty_v8`.

**Pull request merged**
